examples/tv-casting-app/tv-casting-common/core/CastingPlayerDiscovery.cpp

```diff
@@ -0,0 +1,131 @@
+/*
+ *
+ *    Copyright (c) 2023 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "CastingPlayerDiscovery.h"
+
+namespace matter {
+namespace casting {
+namespace core {
+
+using namespace chip::System;
+using namespace chip::Controller;
+using namespace chip::Dnssd;
+using namespace std;
+
+void DeviceDiscoveryDelegateImpl::OnDiscoveredDevice(const chip::Dnssd::DiscoveredNodeData & nodeData)
+{
+
+    VerifyOrReturn(mClientDelegate != nullptr,
+                   ChipLogError(NotSpecified, "CastingPlayerDeviceDiscoveryDelegate, mClientDelegate is a nullptr"));
+
+    // convert nodeData to CastingPlayer
+    Attributes attributes;
+    strcpy(attributes.id, nodeData.resolutionData.hostName);
+
+    char port[kPortMaxLength] = {};
+    snprintf(port, sizeof(port), "%u", nodeData.resolutionData.port);
+    strcat(attributes.id, port);
+
+    strcpy(attributes.name, nodeData.commissionData.deviceName);
+    strcpy(attributes.host_name, nodeData.resolutionData.hostName);
+    attributes.num_IPs = nodeData.resolutionData.numIPs;
+    for (unsigned j = 0; j < attributes.num_IPs; j++)
+    {
+        attributes.ip_address[j] = nodeData.resolutionData.ipAddress[j];
+    }
+    attributes.product_id = nodeData.commissionData.productId;
+    attributes.vendor_id  = nodeData.commissionData.vendorId;
+    attributes.type       = nodeData.commissionData.deviceType;
+
+    Strong<CastingPlayer> player = std::make_shared<CastingPlayer>(attributes);
+
+    std::vector<Strong<CastingPlayer>> * castingPlayers = CastingPlayerDiscovery::GetInstance()->GetCastingPlayers();
+
+    // Add to or update castingPlayers
+    if (castingPlayers->size() != 0)
+    {
+
+        auto it = std::find_if((*castingPlayers).begin(), (*castingPlayers).end(),
+                               [&player](const Strong<CastingPlayer> & castingPlayer) {
+                                   int compareResult = strcmp(castingPlayer->GetId(), player->GetId());
+                                   return (compareResult == 0) ? 1 : 0;
```

Comment:
Why not package this comparison into the CastingPlayer class by overloading operator==()? That would make the comparison implementation reusable elsewhere, instead of it being inline just here. 

See examples here: https://stackoverflow.com/questions/4421706/what-are-the-basic-rules-and-idioms-for-operator-overloading/77055231#77055231

Reply:
updated.